reference/reference.go

```diff
@@ -0,0 +1,191 @@
+package reference
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/docker/distribution/digest"
+	distreference "github.com/docker/distribution/reference"
+	"github.com/docker/docker/image/v1"
+)
+
+const (
+	// DefaultTag defines the default tag used when performing images related actions and no tag or digest is specified
+	DefaultTag = "latest"
+	// DefaultHostname is the default built-in hostname
+	DefaultHostname = "docker.io"
+	// LegacyDefaultHostname is automatically converted to DefaultHostname
+	LegacyDefaultHostname = "index.docker.io"
+	// DefaultRepoPrefix is the prefix used for default repositories in default host
+	DefaultRepoPrefix = "library/"
+)
+
+// Named is an object with a full name
+type Named interface {
+	// Name returns normalized repository name, like "ubuntu".
+	Name() string
```

Comment:
There used to be a `ParseRepositoryInfo` function that would generate the different forms of the name that were needed for different purposes (i.e. `ubuntu`, `library/ubuntu`, `index.docker.io/library/ubuntu`). This PR makes it so that a reference can expose any of these forms using methods. The old code made it very easy to accidentally pass the wrong kind of reference to a function expecting a certain form - for example passing `ubuntu` to something that wants `library/ubuntu`. IMHO this change makes things a lot cleaner by preventing that kind of mistake.